### PR TITLE
Added Repositories tab in Helm Page

### PR DESCRIPTION
**Pull request by lokanandaprabhu**
**JIRA**: 
https://issues.redhat.com/browse/ODC-6715

**Solution Description**: 

- Added Repositories tab in Helm page in developer perspective. 
- This will list down all the namespace and default(All namespaces) repositories. 
- When the logged in user is having access to create either helm chart or project helm chart repository then only the Repository create button will be shown in the action menu in header.
Note: If the disabled property is not there there then it will be default false in the repository details page.

1) If the user don't have create or update access to both HCR and PHCR then the Repositories tab will not be visible and by default it will be Helm Releases page
2) If the user don't have either list or update access for HCR or PHCR then that repository will not be listed down in Repositories tab
3) For the developer user, since they don't have edit access for HCR, HCR's will not be listed down.. only PHCR's can be seen and editable.

**Screen shots / Gifs for design review**: 

Repository Tab in helm

---- For Developer user-----
Aug-01-2022 13-12-50

---For admin user----
admin-helm

**Unit test coverage report**: 
----e2e test results-----
image

**Browser conformance**: 
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

/kind feature

### CODE REVIEW

File: frontend/packages/helm-plugin/src/components/list-page/HelmReleaseListPage.tsx [hunk 1]
```diff
@@ -23,7 +23,7 @@ const PageContents: React.FC<HelmReleaseListPageProps> = (props) => {
   return namespace ? (
     <div>
       <PageHeading title={t('helm-plugin~Helm Releases')} />
-      <HelmReleaseList namespace={namespace} />
+      <HelmReleaseList match={props.match} />
```
Comment (debsmita1) [inline frontend/packages/helm-plugin/src/components/list-page/HelmReleaseListPage.tsx:26]:
```suggestion
      <HelmReleaseList namespace={props.match.params.ns} />
```

File: frontend/packages/helm-plugin/src/utils/serverless-tab-utils.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+export const serverlessTab = (kind: string) => {
```
Comment (debsmita1) [inline frontend/packages/helm-plugin/src/utils/serverless-tab-utils.ts:1]: why serverlessTab in helm ?

**Comment by debsmita1** [comment]
The Repositories list page should have columns
CR name with badge, Display name, namespace, disabled, repo URL , Creation timestamp

File: frontend/packages/helm-plugin/src/components/list-page/HelmReleaseListPage.tsx (hunk 1 above)
Comment (lokanandaprabhu) [reply frontend/packages/helm-plugin/src/components/list-page/HelmReleaseListPage.tsx:26]: HelmReleaseList is also called in HelmTabbedPage.tsx 

There we get match as prop. So I am passing match as prop in HelmReleaseListPage.tsx

File: frontend/packages/helm-plugin/src/utils/serverless-tab-utils.ts (hunk 2 above)
Comment (lokanandaprabhu) [reply frontend/packages/helm-plugin/src/utils/serverless-tab-utils.ts]: Updated the PR. Corrected it.

**Comment by lokanandaprabhu** [comment]
> The Repositories list page should have columns CR name with badge, Display name, namespace, disabled, repo URL , Creation timestamp

Updated the PR.

**Comment by debsmita1** [comment]
We should show the newly added columns on the PHCR & HCR list page when you go from search

File: frontend/packages/helm-plugin/src/components/list-page/HelmChartRepositoryHeader.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      sortField: 'metadata.namespace',
+      transforms: [sortable],
+    },
+    {
+      title: t('helm-plugin~Disabled'),
+      sortField: 'spec.disabled',
+    },
+    {
+      title: t('helm-plugin~Repo URL'),
+      sortField: 'spec.connectionConfig.url',
```
Comment (vikram-raj) [inline frontend/packages/helm-plugin/src/components/list-page/HelmChartRepositoryHeader.tsx:28]: Any reason to not add the `transforms: [sortable]` to `Disabled` and `Repo URL` columns?

File: frontend/packages/helm-plugin/integration-tests/support/pages/helm/helm-details-page.ts [hunk 4]
```diff
@@ -36,4 +36,42 @@ export const helmDetailsPage = {
...
+  },
+  selectHelmTab: (name: string) => {
+    cy.get(`a[data-test-id="horizontal-link-${name}"]`)
+      .should('exist')
+      .click();
+  },
+  verifyHelmActionsDropdown: () => cy.byLegacyTestID('dropdown-button').should('be.visible'),
+  clickHelmActionButton: () => cy.byLegacyTestID('dropdown-button').click(),
+  verifyActionsInCreateMenu: () => {
+    cy.get('.pf-c-dropdown__menu-item')
```
Comment (vikram-raj) [inline frontend/packages/helm-plugin/integration-tests/support/pages/helm/helm-details-page.ts:50]: better to use `data-test-id` to get the dropdown menu item instead of class.

File: frontend/packages/helm-plugin/integration-tests/support/pages/helm/helm-details-page.ts [hunk 5]
```diff
@@ -36,4 +36,42 @@ export const helmDetailsPage = {
...
+      .get('.pf-c-dropdown__menu')
+      .contains(createMenuOption)
+      .click();
+  },
+  clickHelmChartRepository: () => cy.byLegacyTestID('openshift-helm-charts').click(),
+  clickCreateHelmRelease: () => {
+    cy.byLegacyTestID('dropdown-button').click();
+    cy.get('.pf-c-dropdown__menu-item')
+      .contains('Helm Release')
+      .click();
```
Comment (vikram-raj) [inline frontend/packages/helm-plugin/integration-tests/support/pages/helm/helm-details-page.ts:69]: Same here, better to use `data-test-id` instead of class. Or check if you can make use of `selectByDropDownText` function.

File: frontend/packages/helm-plugin/src/components/list-page/HelmChartRepositoryRow.tsx [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import { useTranslation } from 'react-i18next';
+import { TableData } from '@console/internal/components/factory';
+import { ResourceKebab, ResourceLink, Timestamp, Kebab } from '@console/internal/components/utils';
+import { referenceFor, referenceForModel } from '@console/internal/module/k8s';
+import { HelmChartRepositoryModel, ProjectHelmChartRepositoryModel } from '../../models';
+
+const helmChartRepositoryReference = referenceForModel(HelmChartRepositoryModel);
+const projectHelmChartRepositoryReference = referenceForModel(ProjectHelmChartRepositoryModel);
+
+const HelmChartRepositoryRow = ({ obj }) => {
```
Comment (vikram-raj) [inline frontend/packages/helm-plugin/src/components/list-page/HelmChartRepositoryRow.tsx:11]: Define the type for the `HelmChartRepositoryRow` component.

File: frontend/packages/helm-plugin/src/components/list-page/HelmChartRepositoryHeader.tsx (hunk 3 above)
Comment (lokanandaprabhu) [reply frontend/packages/helm-plugin/src/components/list-page/HelmChartRepositoryHeader.tsx]: Updated the PR

File: frontend/packages/helm-plugin/integration-tests/support/pages/helm/helm-details-page.ts (hunk 4 above)
Comment (lokanandaprabhu) [reply frontend/packages/helm-plugin/integration-tests/support/pages/helm/helm-details-page.ts]: Updated the PR

File: frontend/packages/helm-plugin/integration-tests/support/pages/helm/helm-details-page.ts (hunk 5 above)
Comment (lokanandaprabhu) [reply frontend/packages/helm-plugin/integration-tests/support/pages/helm/helm-details-page.ts]: Updated the PR

File: frontend/packages/helm-plugin/src/components/list-page/HelmChartRepositoryRow.tsx (hunk 6 above)
Comment (lokanandaprabhu) [reply frontend/packages/helm-plugin/src/components/list-page/HelmChartRepositoryRow.tsx]: Updated the PR

**Comment by vikram-raj** [comment]
@lokanandaprabhu You need to run `yarn i18n`. Frontend tests are failing due to this.

**Comment by lokanandaprabhu** [comment]
> @lokanandaprabhu You need to run `yarn i18n`. Frontend tests are failing due to this.

I did. Job is triggered again.

File: frontend/packages/helm-plugin/src/components/list-page/RepositoriesRow.tsx [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+          name={obj.metadata.name}
+          namespace={obj.metadata?.namespace}
+        />
+      </TableData>
+      <TableData>{obj.spec?.name ? obj.spec.name : '-'}</TableData>
+      <TableData>
+        {obj.metadata.namespace ? (
+          <ResourceLink kind="Namespace" name={obj.metadata.namespace} />
+        ) : (
+          t('public~All namespaces')
```
Comment (vikram-raj) [inline frontend/packages/helm-plugin/src/components/list-page/RepositoriesRow.tsx:36]:
```suggestion
          t('helm-plugin~All namespaces')
```

**Review by vikram-raj — COMMENTED** [review]
`ProjectHelmChartRepositoryHeader.tsx`, `RepositoriesHeader.tsx` and `HelmChartRepositoryHeader.tsx` are completely same. Can keep one and reuse it at all the places?

**Comment by lokanandaprabhu** [comment]
> `ProjectHelmChartRepositoryHeader.tsx`, `RepositoriesHeader.tsx` and `HelmChartRepositoryHeader.tsx` are completely same. Can keep one and reuse it at all the places?

Used RepositoriesHeader.tsx in all 3 places. Removed PHCR and HCR header files.

**Review by serenamarie125 — CHANGES_REQUESTED** [review]
In Repositories tab, please change "All namespaces" to `All Namespaces` - this will provide consistency with Operator Hub section of the UI

Screen Shot 2022-07-26 at 8 55 39 AM
I

**Comment by serenamarie125** [comment]
Can someone confirm what happens in this case ?  If a non priv user (developer account) goes to Repositories tab, and is able to see the HCR and PHCR in the list view - what happens if they select one of those resources?  

I'm under the impression that they don't have access to view those resources, so understanding what the experience is would be helpful

**Comment by lokanandaprabhu** [comment]
> Can someone confirm what happens in this case ? If a non priv user (developer account) goes to Repositories tab, and is able to see the HCR and PHCR in the list view - what happens if they select one of those resources?
> 
> I'm under the impression that they don't have access to view those resources, so understanding what the experience is would be helpful

To list and to create HCR and PHCR, whether user is having access or not is checked. If the user don't have list access to consider PHCR, then only HCR will be listed. If the user don't have list access for both HCR and PHCR, then not authorised page will be displayed. 

Similarly, if the user has create access to either HCR or PHCR, then only the option to create Repository will be visible.

**Comment by serenamarie125** [comment]
Would like to think this through a bit more.  It sounds like @dperaza4dustbit has had some discussions with Sam Padgett.  If a user doesn't have access to the resources, I don't think we should be showing them.  Which then leads me to this - if a non priv users doesn't have access to either resource type, should we not show the Repositories tab at all?  

Would be good to inventory how we are doing this in other areas, and provide the same experience.

File: frontend/packages/helm-plugin/src/components/list-page/RepositoriesRow.tsx [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        ) : (
+          t('helm-plugin~All Namespaces')
+        )}
+      </TableData>
+      <TableData>{obj.spec?.disabled ? t('helm-plugin~True') : t('helm-plugin~False')}</TableData>
+      <TableData>
+        {obj.spec?.connectionConfig?.url ? (
+          <a href={obj.spec.connectionConfig.url} target="_blank" rel="noopener noreferrer">
+            {obj.spec.connectionConfig.url}
+          </a>
```
Comment (jerolimov) [inline frontend/packages/helm-plugin/src/components/list-page/RepositoriesRow.tsx:44]: Not sure how this would look in this case, but you can try `<ExternalLink` here?

File: frontend/packages/helm-plugin/src/plugin.ts [hunk 9]
```diff
@@ -1,3 +1,20 @@
...
+    type: 'DetailPageBreadCrumbs',
+    properties: {
+      getModels: helmChartRepoModelProviderForBreadcrumbs,
+      breadcrumbsProvider: helmChartRepositoriesBreadcrumbsProvider,
+    },
+    flags: {
+      required: [FLAG_OPENSHIFT_HELM],
+    },
+  },
+];
```
Comment (jerolimov) [inline frontend/packages/helm-plugin/src/plugin.ts:18]: We should add new plugins to the `console-extensions.json` instead. Let me know if this doesn't work for some reason.

File: frontend/packages/helm-plugin/src/providers/useHelmChartRepositoriesBreadcrumbs.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+];
+
+export const useHelmChartRepositoriesBreadcrumbs = (kindObj: K8sKind, urlMatch: match<any>) => {
+  const { t } = useTranslation();
+  const isAdminPerspective = useActivePerspective()[0] === 'admin';
+  return useTabbedTableBreadcrumbsFor(
+    kindObj,
+    urlMatch,
+    'helm-releases',
+    helmPageTab('Repositories'),
```
Comment (jerolimov) [inline frontend/packages/helm-plugin/src/providers/useHelmChartRepositoriesBreadcrumbs.ts:21]: This function was only called one time with `kind === 'Repositories'`. Any reason to not just use a string here?

```suggestion
    'repositories',
```

File: frontend/packages/helm-plugin/src/utils/helm-tab-utils.ts [hunk 11]
```diff
@@ -0,0 +1,8 @@
+export const helmPageTab = (kind: string) => {
```
Comment (jerolimov) [inline frontend/packages/helm-plugin/src/utils/helm-tab-utils.ts:1]: Do we need this small funciton that is only called with one string?

**Comment by debsmita1** [comment]
can you handle this ?
Screenshot 2022-08-01 at 10 50 34 PM

File: frontend/packages/helm-plugin/src/components/list-page/HelmTabbedPage.tsx [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        canCreate,
+      },
+    },
+  ];
+
+  return namespace ? (
+    projectHelmChartCreateAccess ||
+    projectHelmChartEditAccess ||
+    helmChartCreateAccess ||
+    helmChartEditAccess ? (
```
Comment (debsmita1) [inline frontend/packages/helm-plugin/src/components/list-page/HelmTabbedPage.tsx:84]: you might want to check both edit and create accesses for showing the Repositories tab

```suggestion
    (projectHelmChartCreateAccess &&
    projectHelmChartEditAccess )||
    (helmChartCreateAccess &&
    helmChartEditAccess) ? (
```

I created this Role:
```
kind: Role
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: example
  namespace: deb-test
  uid: 1efc50f1-bc8a-4129-86fc-638155839557
  resourceVersion: '699005'
  creationTimestamp: '2022-08-01T17:30:20Z'
  managedFields:
    - manager: Mozilla
      operation: Update
      apiVersion: rbac.authorization.k8s.io/v1
      time: '2022-08-01T17:30:20Z'
      fieldsType: FieldsV1
      fieldsV1:
        'f:rules': {}
rules:
  - verbs:
      - get
      - watch
      - list
      - create
      - delete
    apiGroups:
      - helm.openshift.io
    resources:
      - projecthelmchartrepositories
```
and saw the restricted error in the repositories tab because the user didn't have access to update the chart repository

File: frontend/packages/helm-plugin/src/components/list-page/HelmTabbedPage.tsx [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+      <MultiTabListPage
+        pages={pages}
+        match={props.match}
+        title={t('helm-plugin~Helm')}
+        menuActions={menuActions}
+      />
+    ) : (
+      !loadingCreatePHCR &&
+      !loadingEditPHCR &&
+      !loadingCreateHCR &&
```
Comment (debsmita1) [inline frontend/packages/helm-plugin/src/components/list-page/HelmTabbedPage.tsx:94]: could you explain why is this required?

Comment (lokanandaprabhu) [reply frontend/packages/helm-plugin/src/components/list-page/HelmTabbedPage.tsx:94]: Since we are making call to check access for user, till we get response it will not render the component. If I don't add this, user may experience a glitch if the network is slow. This is just to make sure that till we get response from useAccessReview don't render any component.

**Comment by hemantsaini-7** [comment]
/lgtm

image

**Review by jerolimov — APPROVED** [review]
Looks good so far and has lgtm already :+1:

/approve


Adding docs and px approved based on the Epic ACKs.

/label docs-approved
/label px-approved
